### PR TITLE
feat: Batch automerge save with marking dirty

### PR DESCRIPTION
**Pull request by mykola-vrmchk**
### Motivation / Background

This Pull Request ...

### CODE REVIEW

**Comment by mykola-vrmchk** [comment]
* **#6318** <a href="https://app.graphite.dev/github/pr/dxos/dxos/6318?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @mykola-vrmchk and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts [hunk 1]
```diff
@@ -26,22 +25,23 @@ import {
...
+import { type BeforeSaveParams, LevelDBStorageAdapter } from './leveldb-storage-adapter';
 import { LocalHostNetworkAdapter } from './local-host-network-adapter';
 import { MeshNetworkAdapter } from './mesh-network-adapter';
 import { levelMigration } from './migrations';
-import { type MySublevel } from './types';
+import { type SpaceDoc, type MyLevelBatch, type MySublevel } from './types';
 
 export type { DocumentId };
 
 export interface MetadataMethods {
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts:38]: those need to be renamed

File: packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts [hunk 2]
```diff
@@ -53,13 +53,13 @@ export type AutomergeHostParams = {
 export class AutomergeHost {
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts:53]:
```ts
metadata?: MetadataMethods;
```

`metadata` is not a descriptive name

File: packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts [hunk 3]
```diff
@@ -5,53 +5,67 @@
...
+  constructor(private readonly _params: LevelDBStorageAdapterParams) {
+    super();
+  }
 
   async load(keyArray: StorageKey): Promise<Uint8Array | undefined> {
-    if (this._state !== 'opened') {
-      return undefined;
+    try {
+      if (this._lifecycleState !== LifecycleState.OPEN) {
+        return undefined;
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts:33]: shouldn't this be an error instead

File: packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts [hunk 4]
```diff
@@ -5,53 +5,67 @@
...
+      }
+      return await this._params.db.get<StorageKey, Uint8Array>(keyArray, { ...encodingOptions });
+    } catch (err: any) {
+      // Key not found. Automerge expects undefined in this case.
+      if (err.code !== 'ERR_NOT_FOUND') {
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts:38]: `isLevelDbNotFoundError(err)`

File: packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts [hunk 5]
```diff
@@ -5,53 +5,67 @@
...
+      throw err;
     }
-    return this._params.db
-      .get<StorageKey, Uint8Array>(keyArray, { ...encodingOptions })
-      .catch((err) => (err.code === 'LEVEL_NOT_FOUND' ? undefined : Promise.reject(err)));
   }
 
   async save(keyArray: StorageKey, binary: Uint8Array): Promise<void> {
-    if (this._state !== 'opened') {
+    if (this._lifecycleState !== LifecycleState.OPEN) {
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts:46]: shouldn't this be an error instead

File: packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts [hunk 6]
```diff
@@ -5,53 +5,67 @@
...
     });
+    await batch.write();
+
     await this._params.callbacks?.afterSave?.(keyArray);
   }
 
   async remove(keyArray: StorageKey): Promise<void> {
-    if (this._state !== 'opened') {
+    if (this._lifecycleState !== LifecycleState.OPEN) {
       return undefined;
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts:62]: shouldn't this be an error instead

File: packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts [hunk 7]
```diff
@@ -69,7 +83,7 @@ export class LevelDBStorageAdapter implements StorageAdapterInterface {
   }
 
   async removeRange(keyPrefix: StorageKey): Promise<void> {
-    if (this._state !== 'opened') {
+    if (this._lifecycleState !== LifecycleState.OPEN) {
       return undefined;
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts:87]: shouldn't this be an error instead

File: packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts [hunk 8]
```diff
@@ -5,53 +5,67 @@
...
     }
     await this._params.db.del<StorageKey>(keyArray, { ...encodingOptions });
   }
 
   async loadRange(keyPrefix: StorageKey): Promise<Chunk[]> {
-    if (this._state !== 'opened') {
+    if (this._lifecycleState !== LifecycleState.OPEN) {
       return [];
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/leveldb-storage-adapter.ts:69]: shouldn't this be an error instead

File: packages/core/echo/indexing/src/index-metadata-store.ts [hunk 9]
```diff
@@ -62,15 +62,21 @@ export class IndexMetadataStore implements MetadataMethods {
 
   @trace.span({ showInBrowserTimeline: true })
   @synchronized
-  async markDirty(idToLastHash: Map<string, string>) {
-    const batch = this._lastIndexed.batch();
+  async markDirty(idToLastHash: Map<string, string>, outerBatch?: MyLevelBatch) {
```
Comment (dmaretskyi) [inline packages/core/echo/indexing/src/index-metadata-store.ts:65]: is there a branch where outerBatch is not provided?

File: packages/core/echo/indexing/src/index-metadata-store.ts [hunk 10]
```diff
@@ -62,15 +62,21 @@ export class IndexMetadataStore implements MetadataMethods {
 
   @trace.span({ showInBrowserTimeline: true })
   @synchronized
-  async markDirty(idToLastHash: Map<string, string>) {
-    const batch = this._lastIndexed.batch();
+  async markDirty(idToLastHash: Map<string, string>, outerBatch?: BatchLevel) {
```
Comment (dmaretskyi) [inline packages/core/echo/indexing/src/index-metadata-store.ts:65]: make non-optional

File: packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts [hunk 11]
```diff
@@ -26,22 +24,23 @@ import {
...
 import { MeshNetworkAdapter } from './mesh-network-adapter';
 import { levelMigration } from './migrations';
-import { type MySublevel } from './types';
+import { type SpaceDoc, type BatchLevel, type SubLevelDB } from './types';
 
 export type { DocumentId };
 
 export interface MetadataMethods {
-  markDirty(idToLastHash: Map<string, string>): Promise<void>;
+  markDirty(idToLastHash: Map<string, string>, batch: BatchLevel): Promise<void>;
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts:38]: beforeSave
afterSave

File: packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts [hunk 12]
```diff
@@ -26,22 +24,23 @@ import {
...
+import { type BeforeSaveParams, LevelDBStorageAdapter } from './leveldb-storage-adapter';
 import { LocalHostNetworkAdapter } from './local-host-network-adapter';
 import { MeshNetworkAdapter } from './mesh-network-adapter';
 import { levelMigration } from './migrations';
-import { type MySublevel } from './types';
+import { type SpaceDoc, type BatchLevel, type SubLevelDB } from './types';
 
 export type { DocumentId };
 
 export interface MetadataMethods {
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts:37]: StorageCallbacks

File: packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts (hunk 11 above)
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts:38]: use ContactenedHeadHashes type

File: packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts [hunk 13]
```diff
@@ -26,22 +24,23 @@ import {
...
-  markDirty(idToLastHash: Map<string, string>): Promise<void>;
+  markDirty(idToLastHash: Map<string, string>, batch: BatchLevel): Promise<void>;
+  afterMarkDirty(): Promise<void>;
```
Comment (dmaretskyi) [inline packages/core/echo/echo-pipeline/src/automerge/automerge-host.ts:39]: jsdoc: after leveldb batch commit

File: packages/core/echo/indexing/src/index-store.ts [hunk 14]
```diff
@@ -118,15 +117,13 @@ const headerCodec = {
   },
 };
 
-const getHeader = async (file: File): Promise<{ header: IndexHeader; headerSize: number }> => {
-  const { size } = await file.stat();
-
-  invariant(size > RESERVED_SIZE, 'Index file is too small');
+const getHeader = (data: Buffer): { header: IndexHeader; headerSize: number } => {
```
Comment (dmaretskyi) [inline packages/core/echo/indexing/src/index-store.ts:120]: do we need this if we use level?

**Pull request merged**